File: 020_valid_paranthesis.ts
const isOpening = (s: string) => {
    return s === "{" || s === "[" || s === "(";
};

const isClosing = (s: string) => {
    return s === "}" || s === "]" || s === ")";
};

const isMatching = (s1: string | undefined, s2: string) => {
    return (
        (s1 === "{" && s2 === "}") ||
        (s1 === "[" && s2 === "]") ||
        (s1 === "(" && s2 === ")")
    );
};

function isValid(s: string): boolean {
    const stack: string[] = [];
    for (let i = 0; i < s.length; i++) {
        if (isOpening(s[i])) {
            stack.push(s[i]);
            continue;
        }
        if (!isMatching(stack.pop(), s[i])) {
            return false;
        }
    }
    if (stack.length !== 0) {
        return false;
    }
    return true;
}

File: 8thJune2024/020_valid_paranthesis.cpp
#include <bits/stdc++.h>
using namespace std;

bool is_opening(char ch) { return ch == '(' || ch == '{' || ch == '['; }
bool is_matching(char p1, char p2) {
  return (p1 == '(' && p2 == ')') || (p1 == '[' && p2 == ']') ||
         (p1 == '{' && p2 == '}');
}

class Solution {
public:
  bool isValid(string str) {
    stack<char> s;
    for (char ch : str) {
      if (is_opening(ch)) {
        s.push(ch);
        continue;
      }
      if (s.empty())
        return false;
      if (!is_matching(s.top(), ch)) {
        return false;
      }
      s.pop();
    }
    return s.empty();
  }
};
